shape-detection/resources/endtoend-helpers.js

```diff
@@ -0,0 +1,90 @@
+function imageLoadedPromise(image) {
+    return new Promise(function(resolve, reject) {
+        if (image.complete)
+            resolve();
+        image.addEventListener("load", function() {
+            resolve();
+        });
+    });
+}
+
+function videoLoadedPromise(video) {
+    return new Promise(function(resolve, reject) {
+        video.addEventListener("loadeddata", function() {
+            resolve();
+        });
+    });
+}
+
+function waitForNFrames(n) {
+    return new Promise(function(resolve, reject) {
+        var count = n;
+        function tick() {
+            --count;
+            if (count == 0)
+                resolve();
+            else
+                requestAnimationFrame(tick);
+        }
+        requestAnimationFrame(tick);
+    });
+}
+
+function seekTo(video, time) {
+    return new Promise(function(resolve, reject) {
+        async function handleEvent() {
+            video.removeEventListener("seeked", handleEvent);
+            await waitForNFrames(3);
+            resolve();
+        }
+        video.addEventListener("seeked", handleEvent);
+        video.currentTime = time;
+    });
+}
+
+function checkBoundingBox(actual, expected, fuzziness) {
+    assert_equals(actual.constructor.name, "DOMRectReadOnly");
+    assert_approx_equals(actual.left, expected.left, fuzziness);
+    assert_approx_equals(actual.right, expected.right, fuzziness);
+    assert_approx_equals(actual.top, expected.top, fuzziness);
+    assert_approx_equals(actual.bottom, expected.bottom, fuzziness);
+}
+
+function checkPointsLieWithinBoundingBox(points, boundingBox) {
+    for (point of points) {
+        assert_between_inclusive(point.x, boundingBox.left, boundingBox.right);
+        assert_between_inclusive(point.y, boundingBox.top, boundingBox.bottom);
+    }
+}
+
+function pointIsNear(actual, expected, fuzzinessX, fuzzinessY) {
+    return Math.abs(actual.x - expected.x) <= fuzzinessX && Math.abs(actual.y - expected.y) <= fuzzinessY;
+}
+
+function pointsAreNear(actual, expected, fuzzinessX, fuzzinessY) {
+    for (point of actual) {
+        if (!pointIsNear(point, expected, fuzzinessX, fuzzinessY))
+            return false;
+    }
+    return true;
+}
+
+function average(points) {
+    var averageX = 0;
+    var averageY = 0;
+    var minX = points[0].x;
+    var maxX = points[0].x;
+    var minY = points[0].y;
+    var maxY = points[0].y;
```

Comment:
General comment: globally avoid `var` unless there is a good reason to us it. 

```suggestion
    let averageX = 0;
    let averageY = 0;
    let minX = points[0].x;
    let maxX = points[0].x;
    let minY = points[0].y;
    let maxY = points[0].y;
```

Reply:
I ended up deleting this function.